webhooks/virtualmachine/mutation/scsi_controller.go

```diff
@@ -0,0 +1,388 @@
+	vm *vmopv1.VirtualMachine) (bool, error) {
+
+	var (
+		errs        []error
+		err         error
+		scsiMutated bool
+	)
+	scsiMutated, err = AddSCSIControllersForVolumes(ctx, client, vm)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	// TODO: handle other controller types.
+
+	return scsiMutated, apierrorsutil.NewAggregate(errs)
+}
+
+// AddSCSIControllersForVolumes adds SCSI controllers to the VM as
+// needed based on volume requirements.
+func AddSCSIControllersForVolumes(
+	ctx *pkgctx.WebhookRequestContext,
+	client ctrlclient.Client,
+	vm *vmopv1.VirtualMachine) (bool, error) {
+
+	if len(vm.Spec.Volumes) == 0 {
+		return false, nil
+	}
+
+	// Skip VMs that haven't been created on infrastructure yet. The
+	// VM's spec may not contain controllers from the class/image, and
+	// status won't have device information. The
+	// reconcileSchemaUpgrade method will backfill controllers from
+	// class/image into spec, which will trigger mutation webhook
+	// again.
+	if vm.Status.UniqueID == "" {
+		return false, nil
+	}
+
+	// Initialize hardware spec if needed
+	if vm.Spec.Hardware == nil {
+		vm.Spec.Hardware = &vmopv1.VirtualMachineHardwareSpec{}
+	}
+
+	// Track existing controller configurations
+	existingControllers := make(map[int32]*vmopv1.SCSIControllerSpec)
+	for i := range vm.Spec.Hardware.SCSIControllers {
+		controller := &vm.Spec.Hardware.SCSIControllers[i]
+		existingControllers[controller.BusNumber] = controller
+	}
+
+	// Get device counts from status to check availability
+	statusDeviceCounts := make(map[int32]int32)
+	if vm.Status.Hardware != nil {
+		for _, controller := range vm.Status.Hardware.Controllers {
+			if controller.Type == vmopv1.VirtualControllerTypeSCSI {
+				statusDeviceCounts[controller.BusNumber] = int32(len(controller.Devices)) //nolint:gosec // disable G115
+			}
+		}
+	}
+
+	// Track controllers that need to be added
+	controllersToAdd := make(map[int32]vmopv1.SCSIControllerSpec)
+
+	// Track volume assignments per controller to calculate capacity
+	volumeAssignments := make(map[int32]int32) // busNumber -> count of devices attached.
+
+	// Process each volume to determine controller requirements
+	for i := range vm.Spec.Volumes {
+		vol := &vm.Spec.Volumes[i]
+
+		if vol.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc := vol.PersistentVolumeClaim
+
+		// Determine the target controller based on volume configuration
+		targetController, needsNewController := determineTargetController(
+			pvc,
+			existingControllers,
+			controllersToAdd,
+			statusDeviceCounts,
+			volumeAssignments,
+		)
+
+		if needsNewController {
+			// Check if we can add a new controller
+			if len(existingControllers)+len(controllersToAdd) >= int(maxSCSIControllers) {
+				// Cannot add more controllers. We don't return an
+				// error here because it is possible that by the time
+				// the request reaches the controller, a slot on an
+				// existing controller frees up.
+				continue
```

Comment:
To think out loud, in  this case, we don't fill the busnumber for the volume, and let CSI decide once we pass the BatchAttachment. If there will be any issue attaching the disks, we will have status.volume showing the issue once it got response from BatchAttachment.

This should be acceptable for small scale. For large scale (super large VM), is our use case leaning towards a large number of small VMs? Then yeah this issue is less likely to happen